### PR TITLE
PERF: avoid is_datetime64_any_dtype

### PR DESCRIPTION
**Pull request by jbrockmendel**
This was written with numpy dtypes and DatetimeTZDtype in mind, but will now catch 3rd party and ArrowDtype[timestamp] too, so will be a footgun.

### CODE REVIEW

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -234,6 +234,7 @@ Deprecations
 - Deprecated :func:`is_datetime64tz_dtype`, check ``isinstance(dtype, pd.DatetimeTZDtype)`` instead (:issue:`52607`)
 - Deprecated unused "closed" and "normalize" keywords in the :class:`DatetimeIndex` constructor (:issue:`52628`)
 - Deprecated unused "closed" keyword in the :class:`TimedeltaIndex` constructor (:issue:`52628`)
+- Deprecated :func:`is_datetime64_any_dtype`, check for specific dtypes of interest instead (:issue:`52651`)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply doc/source/whatsnew/v2.1.0.rst:237]: i guess.  all i really care about is that we dont use this internally.  but i do think its a footgun since pretty much anyone who uses it wrote it with only numpy/pd.DatetimeTZDtype in mind

**Comment by jbrockmendel** [comment]
reverted deprecation, so this is just avoiding internal usages

File: pandas/core/dtypes/common.py [hunk 2]
```diff
@@ -1219,7 +1219,18 @@ def is_bool_dtype(arr_or_dtype) -> bool:
 
     if isinstance(arr_or_dtype, ABCIndex):
         # Allow Index[object] that is all-bools or Index["boolean"]
-        return arr_or_dtype.inferred_type == "boolean"
+        if arr_or_dtype.inferred_type == "boolean":
```
Comment (mroeschke) [inline pandas/core/dtypes/common.py:1222]: Hmm this looks unrelated. Did a merge conflict go wrong in this PR?

Comment (jbrockmendel) [reply pandas/core/dtypes/common.py:1222]: agreed, will revert

Comment (jbrockmendel) [reply pandas/core/dtypes/common.py]: Updated + green

File: pandas/core/dtypes/common.py [hunk 3]
```diff
@@ -1219,18 +1219,7 @@ def is_bool_dtype(arr_or_dtype) -> bool:
 
     if isinstance(arr_or_dtype, ABCIndex):
         # Allow Index[object] that is all-bools or Index["boolean"]
-        if arr_or_dtype.inferred_type == "boolean":
```
Comment (mroeschke) [inline pandas/core/dtypes/common.py:1222]: Hmm I thought we wanted to still deprecate this in 2.1?

Comment (jbrockmendel) [reply pandas/core/dtypes/common.py]: un-reverted + green

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel
